Add and use zero!, one!, neg! methods

fingolfin opened the pull request:
Add a bunch of `one!` and `neg!` methods. Also used `zero!`, `one!`, `neg!` in a few more spots. And placed these function in a uniform way into the code based (i.e. these are now always the first three unsafe operations).

This is incomplete in so far as there are still types missing a "native" `neg!` (and in some very few cases also `one!`).

Unfortunately this function is not covered by the ring conformance test suite -- we should change that (in AA).

thofma commented:
Seems to be ok. The unfortunate thing is that we would not notice if a method went missing.

lgoettgens commented on src/flint/fmpq_rel_series.jl:
```diff
@@ -900,6 +900,14 @@ function zero!(z::QQRelPowerSeriesRingElem)
+  ccall((:fmpq_poly_one, libflint), Nothing,
+        (Ref{QQRelPowerSeriesRingElem},), z)
+  z.prec = parent(z).prec_max
+  z.val = parent(z).prec_max
```
this seems wrong. Setting it to one would mean that the valuation is 0 (or I am thinking wrong)
```suggestion
  z.val = 0
```

fingolfin replied:
Yes you are right, this is wrong. Thank you!

lgoettgens commented on src/flint/fmpz_mod.jl:
```diff
@@ -342,6 +342,20 @@ function zero!(z::ZZModRingElem)
+  else
+    z.data = sub!(z.data, R.n, a.data)
+  end
+  return z
```
no idea if this is correct (seems non-trivial enough at least to deserve a comment)

lgoettgens commented on src/flint/fmpz_mod_rel_series.jl:
```diff
@@ -694,6 +694,15 @@ for (etype, rtype, ctype, mtype, brtype, flint_fn) in (
+            (Ref{($etype)}, Ref{($ctype)}),
+            x, x.parent.base_ring.ninv)
+      x.prec = parent(x).prec_max
+      x.val = parent(x).prec_max
```
```suggestion
      x.val = 0
```
see above

lgoettgens commented on src/flint/fmpz_rel_series.jl:
```diff
@@ -591,6 +591,14 @@ function zero!(x::ZZRelPowerSeriesRingElem)
+  ccall((:fmpz_poly_one, libflint), Nothing,
+        (Ref{ZZRelPowerSeriesRingElem},), x)
+  x.prec = parent(x).prec_max
+  x.val = parent(x).prec_max
```
```suggestion
  x.val = 0
```
see above

lgoettgens commented on src/flint/fq_default_rel_series.jl:
```diff
@@ -671,6 +661,22 @@ function zero!(x::FqRelPowerSeriesRingElem)
+  ccall((:fq_default_poly_one, libflint), Nothing,
+        (Ref{FqRelPowerSeriesRingElem}, Ref{FqField}), x, base_ring(x))
+  x.prec = parent(x).prec_max
+  x.val = parent(x).prec_max
```
```suggestion
  x.val = 0
```
see above

lgoettgens commented on src/flint/fq_rel_series.jl:
```diff
@@ -643,6 +635,22 @@ for (etype, rtype, ctype, btype, flint_fn, flint_tail) in (
+      ccall(($(flint_fn*"_one"), libflint), Nothing,
+            (Ref{($etype)}, Ref{($ctype)}), x, base_ring(x))
+      x.prec = parent(x).prec_max
+      x.val = parent(x).prec_max
```
```suggestion
      x.val = 0
```
see above

lgoettgens commented on src/flint/gfp_elem.jl:
```diff
@@ -344,6 +344,15 @@ function zero!(z::fpFieldElem)
 
+function one!(z::fpFieldElem)
+  R = parent(z)
+  return fpFieldElem(UInt(0), R)
```
```suggestion
  return fpFieldElem(UInt(1), R)
```
this looks straight up wrong

lgoettgens reviewed (commented):
found some things. for the valuation issue, can you please check all other relative series types as well (I might have forgotten a comment somewhere).

Overall, I would really like to have https://github.com/Nemocas/AbstractAlgebra.jl/issues/1814 in, but let's merge this once the comments are resolved and fix any issues later

fingolfin commented:
Thanks for the careful feedback, @lgoettgens . I am busy with other things right now, but I plan to go over it all, and maybe use this as incentive to add a few more ring interface tests which hopefully will detect all my mistakes here, to increase confidence in this and further work.

fingolfin commented on src/flint/nmod_rel_series.jl:
```diff
@@ -695,6 +687,23 @@ for (etype, rtype, mtype, brtype, flint_fn) in (
+      ccall(($(flint_fn*"_one"), libflint), Nothing,
+            (Ref{($etype)},), x)
+      x.prec = parent(x).prec_max
+      x.val = parent(x).prec_max
```
```suggestion
      x.val = 0
```

fingolfin commented:
I think this is ready in so far as that I addressed review feedback by @lgoettgens and am not aware of other issues right now. But I'd still feel better if we first had the conformance tests in place to properly check all of these. Thankfully @lgoettgens is working on that.

fingolfin commented:
I made sure this branch passes the new conformance tests in PR #1880  (I.e. I did run the Nemo tests suite with this branch plus the one in PR #1880 merged, against the code in https://github.com/Nemocas/AbstractAlgebra.jl/pull/1827 and it passed).

As such we could also just merge it now.

fingolfin commented on src/flint/nmod.jl:
```diff
@@ -355,6 +355,15 @@ function zero!(z::zzModRingElem)
 
+function one!(z::zzModRingElem)
+  R = parent(z)
+  return zzModRingElem(UInt(1), R)
```
Actually, this one is wrong (as the conformance tests show), and should be
```suggestion
  return zzModRingElem(UInt(R.n == 1 ? 0 : 1), R)
```

But really even better is to just drop this and rely on the default implementation.

lgoettgens replied:
I am in favor of dropping this, but all of the other "dump" implementation right above and below this as well. They are identical to the default, and bring no improvement whatsoever

fingolfin replied:
Yeah I have that locally already but it revealed another bug in AA. PRs will come after lunch.

fingolfin commented:
I've restarted the CI tests to verify the conformance tests pass. They do. I don't see a point in waiting for OscarCI to pass *again*, as they passed before and nothing changed in the code, just in the tests we run.
